### ISSUE TITLE
step function

### ISSUE THREAD

**Issue by EasyArray** [EasyArray/phosphorus#9]
- Maybe build repetitive into the step function?
- And add useful messages (which rules ran, etc.)

**Comment by havijw** [comment]
I can try this. For the messages, I'd need to understand what's happening with the interpret function a little bit better, or just make the step function do everything itself. If the memoization is important though, the latter may not be a good idea. But maybe things aren't so bad because the rules for formal systems don't introduce recursion? In particular, if formal system rules are always O(1) to run, memoizing shouldn't improve the time complexity.

**Comment by havijw** [comment]
I'm almost done with this. One question that came up: when repeating and accumulating, how do you want duplicates in the end result to be handled? There are a couple options:
- At each step, compute using all outputs from all previous steps. End result is the concatenation of all steps computed this way. This is the current behavior, and introduces the most duplicates.
- At each step, compute using only outputs from the previous step. End results is the concatenation of all steps computed this way.
  - This is what I have implemented right now. It still has some duplicates because outputs can be reached by different numbers of steps, but it is faster than the next method because putting together all the lists at the end is _O(n)_, where n is the number of outputs, instead of _O(n<sup>2</sup>)_, which is needed to have no duplicates.
- At each step, compute using only outputs from the previous step. End result is the concatenation of all steps computed this way, but with all duplicates removed.
  - There are no duplicates at the end, but the cost is time efficiency because as each element is being inserted into the final list, the list needs to be checked to see if the element already belongs to it (bringing the complexity to _n<sup>2</sup>_ instead of _n_). This could be done efficiently with a set union, but that destroys the order the outputs were generated, and I think it's nice to have the final list be in that order.

Hopefully this makes sense! I can push the branch and submit a pull request if you want to look at the code.

**Comment by EasyArray** [comment]
I like your choice of ways to handle duplicates -- yes, it's nice to see "historical" duplicates. 

This ties in with the second point, though (showing which rules ran).  I wonder if it would be better to organize the output somewhat (although it would add time complexity). For instance, the output could be a tree with labelled edges, like `{ROOT : MI, I : {MIU}, II :  {MII}}}` or zooming in on a later step: `{ROOT : MIII, I : {ROOT: MIIIU, II : {MIIIUIIIIU}}, II : {ROOT: MIIIIIIII, ...}, III: {{ROOT:MIU, ...}, {ROOT:MUI,...}}`. 

I'm sure there is a cleaner / more efficient way to do this, but take this as an illustration...

**Comment by havijw** [comment]
I like this idea! I'll push what I have right now — showing which rules are applied at each step — but I don't think it would be too hard to parse it into a tree. I think zooming in could just be done by, for example, running the function on `MIII` instead of `MI`.

**Comment by EasyArray** [comment]
Yeah, I didn't mean for zooming in to be an actual feature. Great!
